### ISSUE TITLE
TransportFeedback: long jump due to big reference time

### ISSUE THREAD

**Issue by ibc** [meetecho/janus-gateway#1733]
![ZomboMeme 12082019182116](https://user-images.githubusercontent.com/16191/62880676-37927080-bd2e-11e9-9b15-08f08e55c8c3.jpg)

**Comment by lminiero** [comment]
Ok, I'm back from my vacation, so I can start to care again :stuck_out_tongue_winking_eye: 
Is there an issue we need to fix here too?

**Comment by ibc** [comment]
Yep. Your conversion from 64 or 32 bits to signed 24 is wrong and will produce long jumps that will affect (perhaps just momentarily) to the BWE estimated by the client.

**Comment by atoppi** [comment]
Hi @ibc,
could you please be more specific about this? Which conversion are you referring to?

**Comment by ibc** [comment]
In Transport Feedback generated packets:

If the original 64 bits timestamp is very high, the conversion to **signed** 24 bits (feedback `reference_time`) will produce random positive and negative values for every minor increment of the original timestamp.

We do this and the maths with extra big timestamp values do work:

```c++
// timestamp is uint64_t in ms

this->referenceTime = static_cast<int32_t>((timestamp & 0x1FFFFFC0) / 64);
```

Such a `this->referenceTime` is a `int32_t` value that it's guaranteed to fit into **23** bits (so the highest 24 bit is always 0 and will never be negative when stored in 3 bytes).

Some maths:

```
> timestamp = 8388607 * 64 // timestamp in ms
536870848

> refTime = ((timestamp & 0x1FFFFFC0) / 64)
8388607

> timestamp += 64
536870912

> refTime = ((timestamp & 0x1FFFFFC0) / 64)
0

> timestamp += 64
536870976

> refTime = ((timestamp & 0x1FFFFFC0) / 64)
1
```

And now with the highest timestamp (uint64_t):

```
> timestamp = 2**64 - 1
18446744073709551615

> refTime = ((timestamp & 0x1FFFFFC0) / 64)
8388607

> timestamp += 64
18446744073709551679

> refTime = ((timestamp & 0x1FFFFFC0) / 64)
0
```

**This is:**

* The maximum value in uint64 (18446744073709551615) becomes refTime 8388607 which is 23 bits set to 1.
* When such a highest uint64_t timestamp is increased in 64 it becomes "63" in uint64_t, and refTime becomes 0.
* So in the worst case we jump from 23 bits set to 1 to 24 bits set to 0, which IMHO is ok in a 24 bits Signed value.

Here the problem in Janus:

https://github.com/meetecho/janus-gateway/blob/758286fcadcc7b5d280099579d67f6987425a432/rtcp.c#L1444

What we do to fix it:

```c++
// timestamp is uint64_t

this->referenceTime = static_cast<int32_t>((timestamp & 0x1FFFFFC0) / 64);
```

**Comment by lminiero** [comment]
> What we do to fix it:
> 
> ```c++
> // timestamp is uint64_t
> 
> this->referenceTime = static_cast<int32_t>((timestamp & 0x1FFFFFC0) / 64);
> ```

Ah, thanks for the tip! Which means I'd just need to change that line to this, correct?

	reference_time = (int32_t)((stat->timestamp & 0x1FFFFFC0) / 64);

**Comment by ibc** [comment]
Yes.

**Comment by lminiero** [comment]
Thanks, will commit then :kissing_heart:

**Comment by ibc** [comment]
![Things are getting pretty serious 30082019133335](https://user-images.githubusercontent.com/16191/64017807-d38fea80-cb2a-11e9-80db-70655b07f8d3.jpg)

**Comment by lminiero** [comment]
Discuss stuff like this when I'm not on vacation and I'll jump on it! Otherwise, diff it is :stuck_out_tongue_winking_eye:

**Comment by alexamirante** [comment]
Wait, have you been on vacation? And you're back now? I must have read something about this... well... everywhere! :p

**Comment by lminiero** [comment]
Only in response to issues and pul requests...

**Comment by ibc** [comment]
UOOOOOOO: https://github.com/meetecho/janus-gateway/commit/519dcffa94a4b72e0e52fa8c8ecd51673513ad3a#commitcomment-34895597

**Comment by atoppi** [comment]
Hi @ibc,
this is how the code has been changed:
```
// previous func to calculate reference_time
// reference_time = (stat->timestamp/64000); 
reference_time = (int32_t)(((stat->timestamp / 1000) & 0x1FFFFFC0) / 64);
timestamp = reference_time * 64000;
janus_set3(data, reference_time_pos, reference_time);
```
We are basically converting `reference_time` to a `uint23` (boxed in a `int32`), and as you said its max value becomes `8388607`.
This directly implies that the maximum `timestamp` would be `8388607 * 64000 = 536870848000` (`uint39`).

The problem is that `timestamp` will be compared to `stat->timestamp`:
```
/* Get delta */
if (stat->timestamp>timestamp)
	delta = (stat->timestamp-timestamp)/250;
else
	delta = -(int)((timestamp-stat->timestamp)/250);
/* If it is negative or too big */
if (delta<0 || delta> 127) {
	/* Big one */
	status = janus_rtp_packet_status_largeornegativedelta;
} else {
	/* Small */
	status = janus_rtp_packet_status_smalldelta;
}
```
And `stat->timestamp` is a full `uint64` value stored in the packet.
As a consequence, in high timestamp scenarios, the delta between `stat->timestamp` and calculated `timestamp` will be very high, leading to huge delta feedback.

**TL;DR**
`reference_time` must be a full `uint64` because next computations will consider it like this.
The issue discussed here is caused by `janus_set3` that just copies 3 bytes without considering the destination type (in this case `uint23`).

**Proposed fix**
```
reference_time = (stat->timestamp/64000); 
timestamp = reference_time * 64000;
janus_set3(data, reference_time_pos, reference_time & 0x007FFFFF);
```

I am reviewing the rest of the code, because other signed values might be affected by the same issue.

**Comment by ibc** [comment]
The problem above is that you are storing the `reference_timestamp` as `int23` (as it must be) but then you want to do maths with real `uint64` numbers (those you use to reference "now").

We have solved it by having two separate (but correlated) variables: `referenceTime` and `latestTimestamp`, but created with the same "timestamp" value:

```c++
bool FeedbackRtpTransportPacket::AddPacket(uint16_t sequenceNumber, uint64_t timestamp)
{
	// Let's see if we must set our base.
	if (this->latestTimestamp == 0u)
	{
		MS_DEBUG_DEV("setting base");

		this->baseSequenceNumber   = sequenceNumber + 1;
		this->referenceTime        = static_cast<int32_t>((timestamp & 0x1FFFFFC0) / 64);
		this->latestSequenceNumber = sequenceNumber;
		this->latestTimestamp      = (timestamp >> 6) * 64; // IMPORTANT: Loose precision.

		return true;
	}

	// [...] do stuff

	// Deltas are represented as multiples of 250 us.
	// NOTE: Read it as int 64 to detect long elapsed times.
	int64_t delta64 = (timestamp - this->latestTimestamp) * 4;

	// [...] check delta64 value and do more stuff

	// Update latest seen sequence number and timestamp.
	this->latestSequenceNumber = sequenceNumber;
	this->latestTimestamp      = timestamp;

	return true;
}
```

This is, we store the `referenceTime` as it must finally be (instead of casting it down on write). However we use `lastTimestamp` to compare it with timestamps of next packets.

**Comment by atoppi** [comment]
I see, you are still doing the math on the full `uint64` by using another variable.
So I think our code must be changed to avoid the scenario I described before and so do the math in the right way.

**Comment by ibc** [comment]
And then we must kill the person who decided that having a signed 23 bits number was a good idea.

**Comment by atoppi** [comment]
Actually it is a 24 bit signed integer xD but do not ask me the use cases for negative values

**Comment by ibc** [comment]
Oh yes, 24 bits, sorry.

PS: I've seen Chrome generating, indeed, negative `referenceTime` values in its feedback packets.

**Comment by atoppi** [comment]
I have partially reverted this change and used the bitwise mask to cast `reference_time` on the fly.
Thanks @ibc for the original issue description.

**Comment by hu-ping** [comment]
> The problem above is that you are storing the `reference_time` as `int23` (as it must be) but then you want to do maths with real `uint64` numbers (those you use to reference "now").
> 
> We have solved it by having two separate (but correlated) variables: `referenceTime` and `latestTimestamp`, but created with the same "timestamp" value:
> 
> ```c++
> bool FeedbackRtpTransportPacket::AddPacket(uint16_t sequenceNumber, uint64_t timestamp)
> {
> 	// Let's see if we must set our base.
> 	if (this->latestTimestamp == 0u)
> 	{
> 		MS_DEBUG_DEV("setting base");
> 
> 		this->baseSequenceNumber   = sequenceNumber + 1;
> 		this->referenceTime        = static_cast<int32_t>((timestamp & 0x1FFFFFC0) / 64);
> 		this->latestSequenceNumber = sequenceNumber;
> 		this->latestTimestamp      = (timestamp >> 6) * 64; // IMPORTANT: Loose precision.
> 
> 		return true;
> 	}
> 
> 	// [...] do stuff
> 
> 	// Deltas are represented as multiples of 250 us.
> 	// NOTE: Read it as int 64 to detect long elapsed times.
> 	int64_t delta64 = (timestamp - this->latestTimestamp) * 4;
> 
> 	// [...] check delta64 value and do more stuff
> 
> 	// Update latest seen sequence number and timestamp.
> 	this->latestSequenceNumber = sequenceNumber;
> 	this->latestTimestamp      = timestamp;
> 
> 	return true;
> }
> ```
> 
> This is, we store the `referenceTime` as it must finally be (instead of casting it down on write). However we use `lastTimestamp` to compare it with timestamps of next packets.

hello，Why add 1？this->baseSequenceNumber   = sequenceNumber + 1;

**Comment by ibc** [comment]
> hello，Why add 1？this->baseSequenceNumber = sequenceNumber + 1;

Otherwise first delta would always be 0 which does not make any sense.

BTW which is the purpose of your question?

**Comment by hu-ping** [comment]
> > hello，Why add 1？this->baseSequenceNumber = sequenceNumber + 1;
> 
> Otherwise first delta would always be 0 which does not make any sense.
> 
> BTW which is the purpose of your question?

thank you，in the process of learning mediasoup, I have some doubts about this place.

**Comment by lminiero** [comment]
I think your confusion goes beyond that line of code: this is the Janus repo, not mediasoup :hand_over_mouth:

**Comment by ibc** [comment]
I didn't realize neither 🤣
